api/app/auth.py

```diff
@@ -1,25 +1,29 @@
 """ Authentication module for validating access tokens generated by Keycloak
 """
 import logging
-from fastapi import Depends, HTTPException, status
+from fastapi import Depends, HTTPException, status, Request
 from fastapi.security import OAuth2PasswordBearer
 import jwt
 from app import config
+from app.db.crud.audits import create_audit_log
 
 logger = logging.getLogger(__name__)
 
 # Parse request header and pass the token
 oauth2_scheme = OAuth2PasswordBearer(tokenUrl="/token")
 
 
-async def authenticate(token: str = Depends(oauth2_scheme)):
+async def authenticate(request: Request, token: str = Depends(oauth2_scheme)):
     """ Returns True when validation of the token is successful """
     # RSA public key format
     keycloak_public_key = '-----BEGIN PUBLIC KEY-----\n' + \
         config.get('KEYCLOAK_PUBLIC_KEY') + '\n-----END PUBLIC KEY-----'
 
     try:
-        jwt.decode(token, keycloak_public_key, algorithm='RS256')
+        decoded_token = jwt.decode(token, keycloak_public_key, algorithm='RS256')
+        username = decoded_token['preferred_username']
+        path = request.url.path
+        create_audit_log(username, path)
```

Comment:
I'm busy digging to see if there's a way to separate the audit from the authentication nicely....

Reply:
@conbrad maybe I'm thinking ahead too far - but I suggest we create another function entirely:
`async def audit(request: Request, username: str = Depends(authenticate)):
    """ Audit depends on authenticate """
    logger.info('audit %s for %s', request.url.path, username)
    path = request.url.path
    create_audit_log(username, path)`

and then on a route we can specify like this:
`router = APIRouter(
    prefix="/forecasts",
    dependencies=[Depends(authenticate), Depends(audit)],
)`
That way auth and audit are decoupled a bit better.
Not sure about my proposed function names - one could have a "secure_audit" and an "insecure" audit.